File: bundler-packages/resolutions/src/utils/updatePragma.js
import { fail, succeed } from 'effect/Effect'

export class NoPragmaFoundError extends Error {
	/**
	 * @type {'NoPragmaFoundError'}
	 */
	_tag = 'NoPragmaFoundError'
	/**
	 * @type {'NoPragmaFoundError'}
	 * @override
	 */
	name = 'NoPragmaFoundError'
}

const pragmaPattern = /pragma\s+solidity\s+((\^|~|>|>=|<|<=)?\s*(\d+\.\d+\.\d+)\s*);/
const pragmaPatternWithBounds = /pragma\s+solidity\s+(>=?\d+\.\d+\.\d+)\s*<\s*(\d+\.\d+\.\d+)\s*;/

/**
 * Validates that a Solidity file has a pragma without changing its version constraints.
 * @param {string} solidityCode The Solidity code to update.
 * @param {string} [version] Unused legacy override.
 * @returns {import("effect/Effect").Effect<string, NoPragmaFoundError, never>} The Solidity code.
 */
export const updatePragma = (solidityCode, version) => {
	void version
	let match = solidityCode.match(pragmaPattern)
	if (match?.[3]) {
		return succeed(solidityCode)
	}
	match = solidityCode.match(pragmaPatternWithBounds)
	if (match?.[1]) {
		return succeed(solidityCode)
	}
	return fail(new NoPragmaFoundError('No valid pragma statement found.'))
}

File: resolutions-go/utils/update_pragma.go
package utils

import (
	"fmt"
	"regexp"

	"github.com/williamcory/tevm/go-claude/resolutions-go"
)

var pragmaPattern = regexp.MustCompile(`pragma\s+solidity\s+((\^|~|>|>=|<|<=)?\s*(\d+\.\d+\.\d+)\s*);`)
var pragmaPatternWithBounds = regexp.MustCompile(`pragma\s+solidity\s+(>=?\d+\.\d+\.\d+)\s*<\s*(\d+\.\d+\.\d+)\s*;`)

// UpdatePragma updates the pragma statement in a Solidity file to the specified version
func UpdatePragma(solidityCode string, version string) (string, error) {
	match := pragmaPattern.FindStringSubmatch(solidityCode)
	if len(match) > 3 && match[3] != "" {
		versionToUse := version
		if versionToUse == "" {
			versionToUse = match[3]
		}
		newPragma := fmt.Sprintf("pragma solidity >=%s;", versionToUse)
		return pragmaPattern.ReplaceAllString(solidityCode, newPragma), nil
	}

	match = pragmaPatternWithBounds.FindStringSubmatch(solidityCode)
	if len(match) > 1 && match[1] != "" {
		versionToUse := fmt.Sprintf(">=%s", version)
		if version == "" {
			versionToUse = match[1]
		}
		newPragma := fmt.Sprintf("pragma solidity %s;", versionToUse)
		return pragmaPatternWithBounds.ReplaceAllString(solidityCode, newPragma), nil
	}

	return "", resolutions.NoPragmaFoundError{Message: "No valid pragma statement found."}
}